File: app/src/main/java/com/testapp/testapp/presenter/adapter/VenueAdapter.java
package com.testapp.testapp.presenter.adapter;

import android.content.Intent;
import android.graphics.Color;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.squareup.picasso.Picasso;
import com.testapp.testapp.R;
import com.testapp.testapp.VenueDetailsActivity;
import com.testapp.testapp.model.entity.Photo;
import com.testapp.testapp.model.entity.Venue;
import com.testapp.testapp.model.utils.Constants;
import com.testapp.testapp.model.utils.DistanceFormatter;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by troll on 05.11.2017.
 */

public class VenueAdapter extends CommonRecyclerViewAdapter<Venue, VenueAdapter.Holder> {

    private List<Venue> venues = new ArrayList<>();

    @Override
    public void addItems(List<Venue> items) {
        venues.addAll(items);
        notifyDataSetChanged();
    }

    @Override
    public void clearItems() {
        venues.clear();
        notifyDataSetChanged();
    }

    @Override
    public Holder onCreateViewHolder(ViewGroup parent, int viewType) {
        View row = LayoutInflater.from(parent.getContext()).inflate(R.layout.venue_item_row, parent, false);
        return new Holder(row);
    }

    @Override
    public void onBindViewHolder(Holder holder, int position) {
        Venue currentVenue = venues.get(position);

        holder.setVenueId(currentVenue.getId());
        holder.setLatitude(currentVenue.getLocation().getLatitude());
        holder.setLongitude(currentVenue.getLocation().getLongitude());
        holder.name.setText(currentVenue.getName());
        holder.category.setText(currentVenue.getPrimaryCategory().getName());
        holder.address.setText(currentVenue.getLocation().getAddress());
        if (currentVenue.getPrice() != null)
            holder.price.setText(currentVenue.getPrice().toString());
        else{
            holder.price.setVisibility(View.INVISIBLE);
        }
        if (currentVenue.getLocation().getDistance() != 0.) {
            holder.distance.setText(DistanceFormatter.fromMetersToKilometersFormat(currentVenue.getLocation().getDistance()));
        } else {
            holder.distance.setVisibility(View.GONE);
        }
        if (currentVenue.getRatingColor() != null) {
            holder.rating.setText(String.valueOf(currentVenue.getRating()));
            holder.rating.setBackgroundColor(Color.parseColor("#" + currentVenue.getRatingColor()));
        } else {
            holder.rating.setVisibility(View.INVISIBLE);
        }
        checkVenuePhotos(holder, currentVenue);
    }

    private void checkVenuePhotos(Holder holder, Venue venue){
        if (venue.getPhotoWrapper() == null){
            setVenueImage(holder, venue.getPrimaryCategory().getCategoryIcon().toString());
        } else {
            Photo venuePhoto = venue.getPhotoWrapper().getItems().get(0);
            if (venuePhoto != null){
                venuePhoto.zoomOutWidth();
                venuePhoto.zoomOutHeight();
                setVenueImage(holder, venuePhoto.toString());
            }
        }
    }

    private void setVenueImage(Holder holder, String url){
        Picasso.with(holder.itemView.getContext()).load(url).into(holder.venueImage);
    }

    @Override
    public int getItemCount() {
        return venues.size();
    }

    public class Holder extends RecyclerView.ViewHolder implements View.OnClickListener{

        ImageView venueImage;
        TextView name, category, address, price, distance, rating;

        private String venueId;
        private double latitude, longitude;

        void setLatitude(double latitude) {
            this.latitude = latitude;
        }

        void setLongitude(double longitude) {
            this.longitude = longitude;
        }

        void setVenueId(String venueId) {
            this.venueId = venueId;
        }

        Holder(View itemView) {
            super(itemView);
            itemView.setOnClickListener(this);
            venueImage = (ImageView) itemView.findViewById(R.id.venueImage);
            name = (TextView) itemView.findViewById(R.id.venueName);
            category = (TextView) itemView.findViewById(R.id.venueCategory);
            address = (TextView) itemView.findViewById(R.id.venueAddress);
            price = (TextView) itemView.findViewById(R.id.venuePrice);
            distance = (TextView) itemView.findViewById(R.id.venueDistance);
            rating = (TextView) itemView.findViewById(R.id.venueRating);
        }

        @Override
        public void onClick(View v) {
            Intent intent = new Intent(v.getContext(), VenueDetailsActivity.class);
            String nameForTitle = name.getText().toString();
            intent.putExtra(Constants.INTENT_TITLE, nameForTitle);
            intent.putExtra(Constants.INTENT_VENUE_ID, venueId);
            intent.putExtra(Constants.INTENT_LATITUDE, latitude);
            intent.putExtra(Constants.INTENT_LONGITUDE, longitude);
            v.getContext().startActivity(intent);
        }
    }
}

File: app/src/main/java/com/testapp/testapp/kotlin/presenter/adapter/VenueAdapter.kt
package com.testapp.testapp.kotlin.presenter.adapter

import android.content.Intent
import android.graphics.Color
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import com.squareup.picasso.Picasso
import com.testapp.testapp.KotlinVenueDetailsActivity
import com.testapp.testapp.R
import com.testapp.testapp.kotlin.model.entity.Venue

import com.testapp.testapp.kotlin.System
import com.testapp.testapp.kotlin.model.utils.DistanceFormatter

/**
 * Created by dmitriysmishnyi on 28.08.18.
 */
class VenueAdapter(val venueList: MutableList<Venue>) : CommonRecyclerViewAdapter<Venue, VenueAdapter.Holder>() {
    override fun addItems(items: List<Venue>) {
        venueList.addAll(items)
        notifyDataSetChanged()
    }

    override fun clearItems() {
        venueList.clear()
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup?, viewType: Int): Holder {
        val row = LayoutInflater.from(parent?.context).inflate(R.layout.venue_item_row, parent, false)
        return Holder(row)
    }

    override fun getItemCount(): Int = venueList.size

    override fun onBindViewHolder(holder: Holder?, position: Int) {
        val currentVenue = venueList[position]
        holder?.venueId = currentVenue.id.toInt()
        holder?.latitude = currentVenue.location.latitude
        holder?.longitude = currentVenue.location.longitude
        holder?.name?.text = currentVenue.name
        holder?.category?.text = currentVenue.primaryCategory.name
        holder?.address?.text = currentVenue.location.address
        holder?.price?.text = currentVenue.price.toString()

        if (currentVenue.location.distance != 0) {
            holder?.distance?.text = DistanceFormatter.fromMetersToKilometersFormat(currentVenue.location.distance)
        } else {
            holder?.distance?.visibility = View.GONE
        }

        holder?.rating?.text = currentVenue.rating.toString()
        holder?.rating?.setBackgroundColor(Color.parseColor("#${currentVenue.ratingColor}"))

        Picasso.with(holder?.itemView?.context).load(currentVenue.primaryCategory.icon.toString()).into(holder?.venueImage)
    }

    class Holder(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener {

        var venueId: Int? = null
        var latitude: Double? = null
        var longitude: Double? = null

        val venueImage = itemView.findViewById<ImageView>(R.id.venueImage)!!
        val name = itemView.findViewById<TextView>(R.id.venueName)!!
        val category = itemView.findViewById<TextView>(R.id.venueCategory)!!
        val address = itemView.findViewById<TextView>(R.id.venueAddress)!!
        val price = itemView.findViewById<TextView>(R.id.venuePrice)!!
        val distance = itemView.findViewById<TextView>(R.id.venueDistance)!!
        val rating = itemView.findViewById<TextView>(R.id.venueRating)!!

        init {
            itemView.setOnClickListener(this)
        }

        override fun onClick(v: View?) {
            val intent = Intent(v?.context, KotlinVenueDetailsActivity::class.java)
            val titleName = name.text?.toString()
            intent.putExtra(System.INTENT_TITLE, titleName)
            intent.putExtra(System.INTENT_VENUE_ID, venueId)
            intent.putExtra(System.INTENT_LATITUDE, latitude)
            intent.putExtra(System.INTENT_LONGITUDE, longitude)
            v?.context?.startActivity(intent)
        }
    }
}